core/src/test/scala/org/broadinstitute/dsde/rawls/entities/compact/CompactEntityProviderE2ESpec.scala

```diff
@@ -459,6 +460,85 @@ class CompactEntityProviderE2ESpec extends TestDriverComponentWithFlatSpecAndMat
+    val provider = defaultProvider()
+
+    val metadataBefore = Await.result(provider.entityTypeMetadata(useCache = false, defaultRequestContext), atMost)
+    metadataBefore shouldBe empty
+
+    // create target entities
+    Await.result(provider.createEntity(Entity("targetName1", "targetType", Map()), defaultRequestContext), atMost)
+    Await.result(provider.createEntity(Entity("targetName2", "targetType", Map()), defaultRequestContext), atMost)
+    Await.result(provider.createEntity(Entity("targetName3", "targetType", Map()), defaultRequestContext), atMost)
+    Await.result(provider.createEntity(Entity("targetName4", "targetType", Map()), defaultRequestContext), atMost)
+
+    // create the entity we'll be updating; give it two references
+    Await.result(
+      provider.createEntity(
+        Entity(
+          "sourceName",
+          "sourceType",
+          Map(
+            AttributeName.withDefaultNS("ref1") -> AttributeEntityReference("targetType", "targetName1"),
+            AttributeName.withDefaultNS("ref2") -> AttributeEntityReference("targetType", "targetName2")
+          )
+        ),
+        defaultRequestContext
+      ),
+      atMost
+    )
+
+    // validate the starting references before our batchUpsert
+    val initialReferences =
+      runAndWait(
+        provider.repository.queries.getReferencesFrom(wsid, EntityPointer("sourceType", "sourceName"))
+      )
+
+    initialReferences shouldBe Seq(EntityPointer("targetType", "targetName1"),
+                                   EntityPointer("targetType", "targetName2")
+    )
+
+    // perform the batchUpsert - delete one existing reference, add two more
+    val updates = Source(
+      Seq(
+        EntityUpdateDefinition(
+          "sourceName",
+          "sourceType",
+          Seq(
+            CreateAttributeEntityReferenceList(AttributeName.withDefaultNS("refList")),
+            AddListMember(AttributeName.withDefaultNS("refList"),
+                          AttributeEntityReference("targetType", "targetName3")
+            ),
+            AddListMember(AttributeName.withDefaultNS("refList"), AttributeEntityReference("targetType", "targetName4"))
+          )
+        ),
+        EntityUpdateDefinition("sourceName",
+                               "sourceType",
+                               Seq(
+                                 RemoveAttribute(AttributeName.withDefaultNS("ref2"))
+                               )
+        )
+      )
+    )
+
+    Await.result(provider.batchUpdateEntities(updates, defaultRequestContext), atMost)
+
+    // validate the references after our batchUpsert
+    val finalReferences =
+      runAndWait(
+        provider.repository.queries.getReferencesFrom(wsid, EntityPointer("sourceType", "sourceName"))
+      )
+
+    finalReferences.toSet shouldBe Set(
+      EntityPointer("targetType", "targetName1"),
+      EntityPointer("targetType", "targetName3"),
+      EntityPointer("targetType", "targetName4")
+    )
+
+  }
+
```

Comment:
this new test is almost a total copy of `batchUpsert … update entities with references`, but this one calls `batchUpdate` instead of `batchUpsert`. This test fails on the `develop` branch and verifies the bugfixes in this PR.